### ISSUE TITLE
Transactions

### ISSUE THREAD

**Issue by ivan-kleshnin** [Yomguithereal/baobab#357]
I want to have a sanity checks in derived data (monkeys) to watch for invalid state combinations.

For example, I have a `total` and `pagination` variables which can't be reduced to one variable but are logically coupled and should be consistent with each other.

For example:
```js
// reset pagination
data$.set("total", 0);       // applies immediately and makes state temporary invalid
data$.set("pagination", []); // makes state valid again
```

The question is how to approach this.

In this exact case I can replace two `.set()` calls with one `.merge()`. 
But what about top-level data?

Of course, I don't want to disable `autocommit` feature.

Previously we had `.chain` method to tackle this.
But now:
> You cannot chain update methods now since those will return the affected node's data to better tackle immutability.

Related issue: https://github.com/Yomguithereal/baobab/issues/112

**Comment by Yomguithereal** [comment]
Hello @ivan-kleshnin. The `chain` method was useful when we wouldn't be able to syncwrite the tree. Now that we can, it is probably useless. For your case however, doesn't a `tree.commit()` after both sets the solution?

**Comment by ivan-kleshnin** [comment]
> For your case however, doesn't a tree.commit() after both sets the solution?

Unfortunately, no, because after first write system enters invalid state.

**Comment by Yomguithereal** [comment]
Yes but validation won't run before the end of the transaction and update events also won't fire before this. Do you have some synchronous thing that would mess things up between the first and the second set?

**Comment by ivan-kleshnin** [comment]
> Yes but validation won't run before the end of the transaction and update events also won't fire before this. Do you have some synchronous thing that would mess things up between the first and the second set?

You're right, this shouldn't theoritically happen... But I observed this :confused: 
Let me recheck it again.

**Comment by Yomguithereal** [comment]
Any update on this @ivan-kleshnin?

**Comment by ivan-kleshnin** [comment]
Sorry, I'm temporary out of work. Hope to come back soon.

**Comment by ivan-kleshnin** [comment]
Hello. I'm here at last :)

Yep, something seems to be wrong here. Check this code:

```js
import Baobab from "baobab";

let monkey = Baobab.monkey;

let state = new Baobab({
  models: {
    total: 0,
    pagination: [],

    fullLoad: monkey([
      ["models", "total"],
      ["models", "pagination"],
      function (total, pagination) {
        console.log("fullLoad triggered!");
      }
    ]),
  },
});

let totalCursor = state.select("models", "total");
let paginationCursor = state.select("models", "pagination");

totalCursor.set("1");
paginationCursor.push("245");
paginationCursor.push("246");
```

Now the question. 
**How many times the message `fullLoad triggered!` will be logged?**

#### Expected 

**2** because Baobab collects updates until next event loop sycle. 
So **1** for init and **1** because of set and push.

#### Real 
**3**

---
If we add next line:
```js
paginationCursor.push("2435");
```

Message will be logged one more times. 
So for some reason Baobab does not suspend actions for monkeys.

Since Baobab@2.0.1 it's one trigger less and up to Baobab@2.1.1 behavior is as desribed.

**Comment by Yomguithereal** [comment]
Hello @ivan-kleshnin. Your last sentence is ambiguous: do you mean that with `2.1.1` the library does not behave as it should or the contrary?

Plus, normally, in your example, the message shouldn't be logged at all since you never actually get `/models/check`.

**Comment by ivan-kleshnin** [comment]
> Hello @ivan-kleshnin. Your last sentence is ambiguous: do you mean that with 2.1.1 the library does not behave as it should or the contrary?

Sorry for ambiguity. 
1. I tried this example in 3 versions: **2.0.0**, **2.0.1**, **2.1.1**. 
2. **2.0.1** and **2.1.1** behave the same (as described above). **2.0.0** has even +1 "unnecessary" log
3. I did not try new `lazyMonkey` concept yet.

So, in my tests every version gets extra monkey calls.

> Plus, normally, in your example, the message shouldn't be logged at all since you never actually get /models/check

Right.

**Comment by Yomguithereal** [comment]
@ivan-kleshnin. The bug should now be fixed as per #400. Will release a `2.2.0` very soon.

**Comment by Yomguithereal** [comment]
This bug should be fixed by now.
